Library/src/Multiplayer/MultiplayerConnection.cpp

```diff
@@ -131,27 +131,21 @@ MultiplayerConnection::~MultiplayerConnection()
 	{
 		if (Connected)
 		{
-			const auto _Connection			  = Connection;
-			const auto _WebSocketClient		  = WebSocketClient;
-			const auto _NetworkEventManager	  = NetworkEventManager;
-			const auto _ConversationSystemPtr = ConversationSystemPtr;
+			Poco::Event shutdownEvent;
 
 			DisconnectWithReason("MultiplayerConnection shutting down.",
-								 [_Connection, _WebSocketClient, _NetworkEventManager, _ConversationSystemPtr](ErrorCode)
+								 [&shutdownEvent](ErrorCode)
 								 {
-									 CSP_DELETE(_Connection);
-									 CSP_DELETE(_WebSocketClient);
-									 CSP_DELETE(_NetworkEventManager);
-									 CSP_DELETE(_ConversationSystemPtr);
+									 shutdownEvent.set();
 								 });
+
+			shutdownEvent.wait();
```

Comment:
There's potential for a hang (rather than a hard crash) here if for some reason the callback above is never called. In my (not exhaustive) testing it was always called.

For diagnostic purposes it may be desirable to wait here for some number of milliseconds then log a message indicating it's failing. I would argue against continuing at that point since the program is in an indeterminate state otherwise - this fix was motivated by callbacks being called well after this system had supposedly shut down.